### DISCUSSION TITLE
superset upload csv to clickhosue

### DISCUSSION THREAD

**Discussion by shadowDy** [apache/superset#17256] [Q&A / Help]
Using the upload CSV function of superset, when the table does not exist, the CREATE TABLE statement will be executed.

However, when the linked cluster is a Clickhouse cluster, there is no engine parameter in the statement received by the Clickhouse server, resulting in creation failure and import failure.

Expected create statement:

create table if NOT exists test_ upload(`idc` String,`app` String) ENGINE = MergeTree() PARTITION BY idc ORDER BY (app) SETTINGS index_ granularity = 8192

Statement actually received:

CREATE TABLE test_ upload1 ( "idc app" TEXT ) FORMAT TabSeparatedWithNamesAndTypes

Missing type and engine.

How to solve this problem.

**Comment by rkpotdar** [comment]
Have the same issue. Tried uploading the data from csv into a new clickhouse table but the create fails as the clickhouse expects engine parameter to be sent the in the create query.

Is there a way this can be sent from superset during csv upload?

**Comment by linyuchen** [comment]
Clickhouse create table must set a engine. But the clickhouse_sqlalchemy/drivers/base.py not set a default engine.

```python
    def post_create_table(self, table):
        engine = getattr(table, 'engine', engines.Memory())
        return ' ENGINE = ' + self.process(engine)
```

It works.

**Comment by nanqiangyipo** [comment]
they don't test upload a csv file to clickhouse?

**Comment by 1c7** [comment]
Any updates? I am also trying to upload a CSV file from Superset into a ClickHouse Database

**Comment by mrinterestfull** [comment]
Does this mean we require a patch to 
clickhouse_sqlalchemy/drivers/base.py
?
If yes, can you create it ..or somewhere in params we can pass it:

![image](https://user-images.githubusercontent.com/1559569/225698068-ebd99bcf-fef1-4fb4-b37b-1ec25c37ddea.png)
